src/plugins/workspace/public/components/workspace_form/utils.ts

```diff
@@ -98,3 +115,219 @@ export const convertApplicationsToFeaturesOrGroups = (
     ];
   }, []);
 };
+
+export const isUserOrGroupPermissionSettingDuplicated = (
+  permissionSettings: Array<Partial<WorkspacePermissionSetting>>,
+  permissionSettingToCheck: WorkspacePermissionSetting
+) =>
+  permissionSettings.some(
+    (permissionSetting) =>
+      (permissionSettingToCheck.type === WorkspacePermissionItemType.User &&
+        permissionSetting.type === WorkspacePermissionItemType.User &&
+        permissionSettingToCheck.userId === permissionSetting.userId) ||
+      (permissionSettingToCheck.type === WorkspacePermissionItemType.Group &&
+        permissionSetting.type === WorkspacePermissionItemType.Group &&
+        permissionSettingToCheck.group === permissionSetting.group)
+  );
+
+// default permission mode is read
+export const getPermissionModeId = (modes: WorkspacePermissionMode[]) => {
+  for (const key in optionIdToWorkspacePermissionModesMap) {
+    if (optionIdToWorkspacePermissionModesMap[key].every((mode) => modes?.includes(mode))) {
+      return key;
+    }
+  }
+  return PermissionModeId.Read;
+};
+
+export const convertPermissionSettingsToPermissions = (
+  permissionItems: WorkspacePermissionSetting[] | undefined
+) => {
+  if (!permissionItems || permissionItems.length === 0) {
+    return undefined;
+  }
+  return permissionItems.reduce<SavedObjectPermissions>((previous, current) => {
+    current.modes.forEach((mode) => {
+      if (!previous[mode]) {
+        previous[mode] = {};
+      }
+      switch (current.type) {
+        case 'user':
+          previous[mode].users = previous[mode].users?.includes(current.userId)
+            ? previous[mode].users
+            : [...(previous[mode].users || []), current.userId];
+          break;
+        case 'group':
+          previous[mode].groups = previous[mode].groups?.includes(current.group)
+            ? previous[mode].groups
+            : [...(previous[mode].groups || []), current.group];
+          break;
+      }
+    });
+    return previous;
+  }, {});
+};
+
+const isWorkspacePermissionMode = (test: string): test is WorkspacePermissionMode =>
+  test === WorkspacePermissionMode.LibraryRead ||
+  test === WorkspacePermissionMode.LibraryWrite ||
+  test === WorkspacePermissionMode.Read ||
+  test === WorkspacePermissionMode.Write;
+
+export const convertPermissionsToPermissionSettings = (permissions: SavedObjectPermissions) => {
```

Comment:
convertPermissionsToPermissionSettings and convertPermissionSettingsToPermissions kind of similarity in iterating over the permissionItems or permissions array and processing each item, is there any way to improve here?

Reply:
I think it's a little bit hard to separate similar logic from them. The data structure of permissions is different from permissionSettings.
```ts
// permissions
interface SavedObjectPermissions{
    library_read: {
        users: [],
        groups: [],
    },
    library_write: {
        users: [],
        groups: [],
    }
    read: {
        users: [],
        groups: [],
    },
    write: {
        users: [],
        groups: [],
    }
}

// permissionSettings
export type WorkspacePermissionSetting = {
    id: number;
    type: 'user';
    modes: WorkspacePermissionMode;
    userId: string;
} | {
    id: number;
    type: 'group';
    modes: WorkspacePermissionMode;
    group: string;
}
```
`convertPermissionsToPermissionSettings` has similar logic for processing users or groups. We can optimize this part.